src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.cs

```diff
@@ -0,0 +1,304 @@
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        internal static DiagnosticDescriptor PreferStreamWriteAsyncMemoryOverloadsRule = DiagnosticDescriptorHelper.Create(
+                                                                                        RuleId,
+                                                                                        s_localizableTitleWrite,
+                                                                                        s_localizableMessageWrite,
+                                                                                        DiagnosticCategory.Performance,
+                                                                                        RuleLevel.IdeSuggestion,
+                                                                                        s_localizableDescriptionWrite,
+                                                                                        isPortedFxCopRule: false,
+                                                                                        isDataflowRule: false);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+            ImmutableArray.Create(PreferStreamReadAsyncMemoryOverloadsRule, PreferStreamWriteAsyncMemoryOverloadsRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(AnalyzeCompilationStart);
+        }
+
+        private void AnalyzeCompilationStart(CompilationStartAnalysisContext context)
+        {
+            // Find the essential type for this analysis
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemIOStream, out INamedTypeSymbol? streamType))
+            {
+                return;
+            }
+
+            // Find the types for the rule message, available since .NET Standard 2.1
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemReadOnlyMemory1, out INamedTypeSymbol? readOnlyMemoryType))
+            {
+                return;
+            }
+            if (!context.Compilation.TryGetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemMemory1, out INamedTypeSymbol? memoryType))
+            {
+                return;
+            }
+
+            // Retrieve the 3+ ReadAsync and WriteSync methods available in Stream
+            // If we find fewer than 3, the Memory based overloads are not supported in this .NET version
+            IEnumerable<IMethodSymbol> readAsyncMethodGroup = streamType.GetMembers("ReadAsync").OfType<IMethodSymbol>();
+            if (readAsyncMethodGroup.Count() < 3)
```

Comment:
I see that you are actually fetching each undesired and preferred method symbol, and bailing out if any of these are missing. Do we really need the count checks here?

Reply:
I was trying to quit early here: I am not sure what is the probability of _not_ finding the newest overloads (users of versions lower than .NET Standard 2.1), but I was doing the early check of `Count() < 3` for those users. This would've also worked if in the future we introduce new overloads.
Without the `Count()` checks, then we will have to retrieve the types of the parameters for those methods: the cancellation token type, the byte special type, the int special type, and the actual retrieval of the preferred/undesired methods begins at line 227.
I'll remove the `Count()` calls, but do let me know if you would rather keep them.

Reply:
I think we can remove the Count() calls - they will become redundant over time as everyone migrates forward.

Reply:
Removed.